### PR TITLE
test: add tests for associated type `Eff`

### PR DESCRIPTION
**Pull request by jaschdoc**
- [x] Add instance for Int32.
- [x] Add for String
- [x] Add instance for Vector[a]
- [x] Write polymorphic functions that use the trait.
- [x] Write monomorphic tests that exercise the trait directly and also the polymorphic functions.
- [x] Add tests
- [x] Reduce crash example below

Related to https://github.com/flix/flix/issues/7424

### CODE REVIEW

**Comment by jaschdoc** [comment]
I will try to reduce this example later :)

```
mod Test.Dec.Assoc.Type.Eff {

    ////////////////////////////////////////////////////////
    // Type Definitions                                   //
    ////////////////////////////////////////////////////////

    eff OutInt32 {
        pub def toStream(x: Int32): Unit
    }

    eff OutString {
        pub def toStream(x: String): Unit
    }

    trait Runner[a] {
        pub type E: Eff
        pub def run(x: a): Unit \ Runner.E[a]
    }

    instance Runner[Int32] {
        pub type E = OutInt32
        pub def run(x: Int32): Unit \ OutInt32 =
            do OutInt32.toStream(x)
    }

    instance Runner[String] {
        pub type E = OutString
        pub def run(x: String): Unit \ OutString =
            do OutString.toStream(x)
    }

    instance Runner[Vector[a]] with Runner[a] {
        pub type E = Runner.E[a]
        pub def run(x: Vector[a]): Unit \ Runner.E[a] =
            foreach (a <- x) Runner.run(a)
    }

    ////////////////////////////////////////////////////////
    // Polymorphic functions using Runner                 //
    ////////////////////////////////////////////////////////

    pub def runOnce(x: a): Unit \ Runner.E[a] with Runner[a] =
        Runner.run(x)

    pub def runTwice(x: a): Unit \ Runner.E[a] with Runner[a] = {
        Runner.run(x);
        Runner.run(x)
    }


    ////////////////////////////////////////////////////////
    // Monomorphic functions using Runner                 //
    ////////////////////////////////////////////////////////

    // Int32

    pub def runOnceInt3201(x: Int32): Unit \ OutInt32 =
        Runner.run(x)

    pub def runTwiceInt3201(x: Int32): Unit \ OutInt32 = {
        Runner.run(x);
        Runner.run(x)
    }

    pub def runOnceInt3202(x: Int32): Unit \ OutInt32 =
        do OutInt32.toStream(x)

    pub def runTwiceInt3202(x: Int32): Unit \ OutInt32 = {
        do OutInt32.toStream(x);
        do OutInt32.toStream(x)
    }

    // String

    pub def runOnceString01(x: String): Unit \ OutString =
        Runner.run(x)

    pub def runTwiceString01(x: String): Unit \ OutString = {
        Runner.run(x);
        Runner.run(x)
    }

    pub def runOnceString02(x: String): Unit \ OutString =
        do OutString.toStream(x)

    pub def runTwiceString02(x: String): Unit \ OutString = {
        do OutString.toStream(x);
        do OutString.toStream(x)
    }

    // Vector[Int32]

    pub def runOnceVectorInt3201(x: Vector[Int32]): Unit \ OutInt32[Int32] = // Error occurred here trying to change effect from Runner.E[Int32] to OutInt32 manually
        Runner.run(x)

}
```

```
#
# An unexpected error has been detected by the Flix compiler:
#
#   Unexpected type: 'Error'.
#
# This is a bug in the Flix compiler. Please report it here:
#
# https://github.com/flix/flix/issues
#
# -- Flix Compiler --
#
# Flix Version : 0.45.0
#   incremental: All
#
# -- Java Virtual Machine --
#
# JVM Version  : 21.0.1 (2023-10-17)
# JVM Vendor   : GraalVM Community
# JAVA_HOME    : /home/_/.sdkman/candidates/java/21.0.1-graalce
# System       : Linux
#
# -- Stack Trace --
ca.uwaterloo.flix.util.InternalCompilerException: Unexpected type: 'Error'.
	at ca.uwaterloo.flix.language.phase.unification.BoolAlg.fromType(BoolAlg.scala:157)
	at ca.uwaterloo.flix.language.phase.unification.BoolAlg.fromType$(BoolAlg.scala:142)
	at ca.uwaterloo.flix.language.phase.unification.BoolFormulaAlg.fromType(BoolFormulaAlg.scala:28)
	at ca.uwaterloo.flix.language.phase.unification.EffUnification$.lookupOrSolve(EffUnification.scala:243)
	at ca.uwaterloo.flix.language.phase.unification.EffUnification$.unify(EffUnification.scala:171)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.$anonfun$resolveEquality$4(ConstraintSolver.scala:309)
	at ca.uwaterloo.flix.util.Result.flatMap(Result.scala:62)
	at ca.uwaterloo.flix.util.Result.flatMap$(Result.scala:61)
	at ca.uwaterloo.flix.util.Result$Ok.flatMap(Result.scala:94)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.$anonfun$resolveEquality$2(ConstraintSolver.scala:308)
	at ca.uwaterloo.flix.util.Result.flatMap(Result.scala:62)
	at ca.uwaterloo.flix.util.Result.flatMap$(Result.scala:61)
	at ca.uwaterloo.flix.util.Result$Ok.flatMap(Result.scala:94)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolveEquality(ConstraintSolver.scala:307)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolveOne(ConstraintSolver.scala:271)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.tryResolve$1(ConstraintSolver.scala:250)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolveOneOf(ConstraintSolver.scala:260)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.resolve(ConstraintSolver.scala:226)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.visitSpec(ConstraintSolver.scala:110)
	at ca.uwaterloo.flix.language.phase.typer.ConstraintSolver$.visitDef(ConstraintSolver.scala:56)
	at ca.uwaterloo.flix.language.phase.Typer$.visitDef(Typer.scala:177)
	at ca.uwaterloo.flix.language.phase.Typer$.$anonfun$visitDefs$2(Typer.scala:162)
	at ca.uwaterloo.flix.util.ParOps$.$anonfun$parTraverseValues$1(ParOps.scala:106)
	at ca.uwaterloo.flix.util.ParOps$.$anonfun$parMap$2(ParOps.scala:64)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)


```

**Comment by magnus-madsen** [comment]
@mlutze Lets both take a look?

File: main/test/flix/Test.Dec.Assoc.Type.Eff.flix [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+
+    @Test
+    pub def testString04(): Bool =
+        handleWithListString(runTwiceString02, "Hello") == "Hello" :: "Hello" :: Nil
+
+
+    // Vector[Int32]
+
+    @Test
+    pub def testVectorInt3201(): Bool =
```
Comment (mlutze) [inline main/test/flix/Test.Dec.Assoc.Type.Eff.flix:225]: Can we make this Int32_01?

**Comment by magnus-madsen** [comment]
LGMT

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**